x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -969,6 +966,71 @@ public void logRemovedField(String parserName, Supplier<XContentLocation> locati
         }
     }
 
+    private void doUpdateApiKey(
+        final Authentication authentication,
+        final UpdateApiKeyRequest request,
+        final Set<RoleDescriptor> userRoles,
+        final VersionedApiKeyDocWithSource currentVersionedDoc,
+        final ActionListener<UpdateApiKeyResponse> listener
+    ) throws IOException {
+        logger.trace(
+            "Building update request for API key doc [{}] with seqNo [{}] and primaryTerm [{}]",
+            request.getId(),
+            currentVersionedDoc.seqNo(),
+            currentVersionedDoc.primaryTerm()
+        );
+        final var currentDocVersion = Version.fromId(currentVersionedDoc.doc().version);
+        final var targetDocVersion = clusterService.state().nodes().getMinNodeVersion();
+        assert currentDocVersion.onOrBefore(targetDocVersion) : "current API key doc version must be on or before target version";
+        if (currentDocVersion.before(targetDocVersion)) {
+            logger.debug(
+                "API key update for [{}] will update version from [{}] to [{}]",
+                request.getId(),
+                currentDocVersion,
+                targetDocVersion
+            );
+        }
+
+        final IndexRequest indexRequest = client.prepareIndex(SECURITY_MAIN_ALIAS)
+            .setId(request.getId())
+            .setSource(
+                buildUpdatedDocument(
+                    currentVersionedDoc.doc(),
+                    authentication,
+                    userRoles,
+                    request.getRoleDescriptors(),
+                    targetDocVersion,
+                    request.getMetadata()
+                )
+            )
+            .setIfSeqNo(currentVersionedDoc.seqNo())
+            .setIfPrimaryTerm(currentVersionedDoc.primaryTerm())
+            .setOpType(DocWriteRequest.OpType.INDEX)
+            .request();
+
+        final boolean noop = indexRequest.source().equals(currentVersionedDoc.source());
```

Comment:
I'm going for this approach for a noop check because it's very simple, and robust. If we change what fields can be updated or evolve the API key doc model, we can still rely on the noop check working here.

We could for sure optimize this (for instance by weaving the noop check into `buildUpdatedDocument`) but my intuition is that would be premature. Could likewise avoid passing around the source for the doc with the doc (it's redundant logically and on the heap) but then we'd have some form of conversion here.

I'm open to changing this but wanted to get your thoughts first.

Reply:
> it's very simple, and robust

It's simple. But I don't think it's robust. This relies on the order of `buildUpdatedDocument` to match exactly the index mapping for API Key document. They are probably identical right now. But we never intend for them to be consistent. There is also no guarantee that no optimisation on lucene side could re-arrange how bytes are read from the disc. Overall I think matching these two (potentially big) string blobs for equality is not desirable.